integration/delete_command_compat_test.go

```diff
@@ -0,0 +1,150 @@
+// limitations under the License.
+
+package integration
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.mongodb.org/mongo-driver/bson"
+
+	"github.com/FerretDB/FerretDB/integration/setup"
+	"github.com/FerretDB/FerretDB/integration/shareddata"
+)
+
+// deleteCommandCompatTestCase describes delete compatibility test case.
+type deleteCommandCompatTestCase struct {
+	skip    string // optional, skip test with a specified reason
+	deletes bson.A // required
+}
+
+func testDeleteCommandCompat(t *testing.T, testCases map[string]deleteCommandCompatTestCase) {
+	t.Helper()
+
+	for name, tc := range testCases {
+		name, tc := name, tc
+		t.Run(name, func(t *testing.T) {
+			t.Helper()
+
+			if tc.skip != "" {
+				t.Skip(tc.skip)
+			}
+
+			t.Parallel()
+
+			// Use per-test setup because deletes modify data set.
+			ctx, targetCollections, compatCollections := setup.SetupCompat(t)
+
+			for i := range targetCollections {
+				targetCollection := targetCollections[i]
+				compatCollection := compatCollections[i]
+				t.Run(targetCollection.Name(), func(t *testing.T) {
+					t.Helper()
+
+					var targetRes, compatRes bson.D
+					targetErr := targetCollection.Database().RunCommand(ctx,
+						bson.D{
+							{"delete", targetCollection.Name()},
+							{"deletes", tc.deletes},
+						}).Decode(&targetRes)
+					compatErr := compatCollection.Database().RunCommand(ctx,
+						bson.D{
+							{"delete", compatCollection.Name()},
+							{"deletes", tc.deletes},
+						}).Decode(&compatRes)
+
+					if targetErr != nil {
+						t.Logf("Target error: %v", targetErr)
+						targetErr = UnsetRaw(t, targetErr)
+						compatErr = UnsetRaw(t, compatErr)
+						assert.Equal(t, compatErr, targetErr)
+					} else {
+						require.NoError(t, compatErr, "compat error; target returned no error")
+					}
+
+					assert.Equal(t, compatRes, targetRes)
+
+					targetDocs := FindAll(t, ctx, targetCollection)
+					compatDocs := FindAll(t, ctx, compatCollection)
+
+					t.Logf("Compat (expected) IDs: %v", CollectIDs(t, compatDocs))
+					t.Logf("Target (actual)   IDs: %v", CollectIDs(t, targetDocs))
+					AssertEqualDocumentsSlice(t, compatDocs, targetDocs)
+				})
+			}
+		})
+	}
+}
+
+func TestDeleteCommandCompat(t *testing.T) {
+	t.Parallel()
+
+	testCases := map[string]deleteCommandCompatTestCase{
+		"OneLimited": {
+			deletes: bson.A{
+				bson.D{{"q", bson.D{{"v", int32(0)}}}, {"limit", 1}},
+			},
+		},
+		"TwoLimited": {
+			deletes: bson.A{
+				bson.D{{"q", bson.D{{"v", int32(42)}}}, {"limit", 1}},
+				bson.D{{"q", bson.D{{"v", "foo"}}}, {"limit", 1}},
+			},
+			skip: "https://github.com/FerretDB/FerretDB/issues/2935",
+		},
+		"DuplicateFilter": {
+			deletes: bson.A{
+				bson.D{{"q", bson.D{{"v", "foo"}}}, {"limit", 1}},
+				bson.D{{"q", bson.D{{"v", "foo"}}}, {"limit", 1}},
+			},
+			skip: "https://github.com/FerretDB/FerretDB/issues/2935",
```

Comment:
I don't understand that. If MongoDB always behaves differently (as the issue description implies), we should fix out code, not this test (as the issue title implies). If it behaves differently sometimes due to ordering of documents in collection, then there is nothing we could do, and then there is no need to keep that test and issue we could never "fix".

Reply:
Yes, the issue description is not correct, we should fix the code.